File: clients/python/game.py
import time
import client_json
from ai import AI
import json
import sys
import game_objects
import operator
import utility
import socket


class GameOverException(Exception):
    def __init__(self, winner, reason):
        Exception.__init__(self)
        self.winner = winner
        self.reason = reason

class Game:

    def __init__(self, conn, addr, port, name):
        self.serv_conn = conn
        self.serv_addr = addr
        self.serv_port = port
        self.game_name = name
        self.ai = AI()
        self.ai.connection = self.serv_conn

    #Attempt to connect to the server
    def connect(self):
        while True:
            try:
                #Attempting to connect
                self.serv_conn.connect((self.serv_addr, self.serv_port))
            except socket.error:
                #Failed to connect
                time.sleep(1)
            else:
                #Client connected
                return True

    def receive(self):
        data = utility.receive_string(self.serv_conn)
        message = json.loads(data)

        if message['type'] == 'changes':
            self.update_game(message)
        elif message['type'] == 'player_id':
            self.ai.my_player_id = message['args']['id']
        elif message['type'] == 'game_over':
            raise GameOverException(message["args"]["winner"], message["args"]["reason"])
        return message

    def wait_for(self, *types):
        while True:
            message = self.receive()
            if message['type'] in types:
                return message

    #Attempt to login to the server
    def login(self):
        login_json = client_json.login.copy()
        login_json['args']['username'] = self.ai.username
        login_json['args']['password'] = self.ai.password

        utility.send_string(self.serv_conn, json.dumps(login_json))

        message = self.wait_for('success', 'failure')
        if message['type'] == 'success':
            #Login success
            return True
        else:
            #Login failed
            return False

    #Attempt to create a game on the server
    def create_game(self):
        create_game_json = client_json.create_game.copy()
        if self.game_name is not None:
            create_game_json['args']['game_name'] = self.game_name

        utility.send_string(self.serv_conn, json.dumps(create_game_json))

        message = self.wait_for('success', 'failure')
        if message['type'] == "success":
            self.game_name = message['args']['name']
            print("Game created: {}".format(self.game_name))
            return True
        else:
            #Game creation failed
            return False

    #Receive Player ID from server
    def recv_player_id(self):
        self.wait_for('player_id')
        return True

    #Runs before main_loop has began.
    def init_main(self):
        self.wait_for('start_game')

        self.ai.init()
        return True

    #Runs after main_loop has finished.
    def end_main(self):
        self.ai.end()
        return True

    #Main connection loop until end of game.
    def main_loop(self):
        while True:
            message = self.wait_for('start_turn', 'game_over')
            if message['type'] == 'game_over':
                return True

            if self.ai.my_player_id == self.ai.player_id:
                utility.v_print("Turn Number: {}".format(self.ai.turn_number))
                self.ai.run()
                utility.send_string(self.serv_conn, json.dumps(client_json.end_turn))

     
    def get_log(self):
        log_json = client_json.get_log.copy()
        utility.send_string(self.serv_conn, json.dumps(log_json))

        message = self.wait_for('success', 'failure')
        if message['type'] == "success":
            file = open(self.game_name + '.glog', 'wb')
            file.write(message['args']['log'].encode('utf-8'))
            file.close()

    #Update game from message
    def update_game(self, message):
        if message.get("type") != "changes":
            return False

        for change in message.get("args").get("changes"):
            if change.get("action") == "add":
                self.change_add(change)

            elif change.get("action") == "remove":
                self.change_remove(change)

            elif change.get("action") == "update":
                self.change_update(change)

            elif change.get("action") == "global_update":
                self.change_global_update(change)

        return True

    #Parse the add action
    def change_add(self, change):
        values = change.get("values")
        if change.get("type") == "Player":
            temp = game_objects.Player(connection=self.serv_conn, parent_game=self, id=values.get("id"), name=values.get("name"), byte_dollars=values.get("byte_dollars"), cycles=values.get("cycles"), time=values.get("time"))
            self.ai.players.append(temp)
        if change.get("type") == "Virus":
            temp = game_objects.Virus(connection=self.serv_conn, parent_game=self, id=values.get("id"), x=values.get("x"), y=values.get("y"), owner=values.get("owner"), level=values.get("level"), moves_left=values.get("moves_left"), living=values.get("living"))
            self.ai.viruses.append(temp)
        if change.get("type") == "Base":
            temp = game_objects.Base(connection=self.serv_conn, parent_game=self, id=values.get("id"), x=values.get("x"), y=values.get("y"), owner=values.get("owner"), spawns_left=values.get("spawns_left"))
            self.ai.bases.append(temp)
        if change.get("type") == "Tile":
            temp = game_objects.Tile(connection=self.serv_conn, parent_game=self, id=values.get("id"), x=values.get("x"), y=values.get("y"), owner=values.get("owner"))
            self.ai.tiles.append(temp)
        return True

    #Parse the remove action.
    def change_remove(self, change):
        remove_id = change.get("id")
        for player in self.ai.players:
            if player.id == remove_id:
                self.ai.players.remove(player)
                return True
        for virus in self.ai.viruses:
            if virus.id == remove_id:
                self.ai.viruses.remove(virus)
                return True
        for base in self.ai.bases:
            if base.id == remove_id:
                self.ai.bases.remove(base)
                return True
        for tile in self.ai.tiles:
            if tile.id == remove_id:
                self.ai.tiles.remove(tile)
                return True
        return False

    #Parse the update action.
    def change_update(self, change):
        change_id = change.get("id")
        values = change.get("values")
        for player in self.ai.players:
            if player.id == change_id:
                player.__dict__.update(values)
                return True
        for virus in self.ai.viruses:
            if virus.id == change_id:
                virus.__dict__.update(values)
                return True
        for base in self.ai.bases:
            if base.id == change_id:
                base.__dict__.update(values)
                return True
        for tile in self.ai.tiles:
            if tile.id == change_id:
                tile.__dict__.update(values)
                return True
        return False

    #Parse the global_update action
    def change_global_update(self, change):
        values = change.get("values")
        self.ai.__dict__.update(values)
        return True

    def run(self):
        if not self.connect(): return False
        if not self.login(): return False
        if not self.create_game(): return False
        if not self.recv_player_id(): return False

        if not self.init_main(): return False
        try:
            self.main_loop()
        except GameOverException as e:
            if e.winner == self.ai.my_player_id:
                game_over_message = "You Win! - {reason}".format(reason=e.reason)
            else:
                game_over_message = "You Lose! - {reason}".format(reason=e.reason)
        else:
            game_over_message = "Game over was never reached."

        if not self.end_main(): return False
        print(game_over_message)

        if not self.get_log(): return False
        

File: clients/cpp/Game.cpp
#include "Game.h"
#include "json/json.h"
#include <iostream>
#include <fstream>
#include <sstream>

Game::Game(GameSocket& conn, std::string addr, int port, std::string name,
           int verbosity2)
{
    this->conn = conn;
    this->addr = addr;
    this->port = port;
    this->name = name;
    ai.connection = conn;
    verbosity = verbosity2;
}

bool Game::connect()
{
    return conn.open_server_connection(addr,port);
}

std::string Game::receive()
{
    std::string message = conn.rec_string();
    if(verbosity == 2)
    {
        std::cout<<"Recieved: "<<message<<std::endl;
    }
    Json::Value root;
    Json::Reader reader;
    reader.parse(message,root,false);
    if(root["type"].asString() == "changes")
    {
        update_game(message);
    }
    else if(root["type"].asString() == "player_id")
    {
        ai.my_player_id = root["args"]["id"].asInt();
    }
    else if(root["type"].asString() == "game_over")
    {
        throw GameOverException(root["args"]["winner"].asInt(),
                                root["args"]["reason"].asString());
    }
    return message;
}

std::string Game::wait_for(std::vector<std::string>& types)
{
    while(true)
    {
        std::string message = receive();
        Json::Value root;
        Json::Reader reader;
        reader.parse(message,root,false);
        for(int i = 0; i < types.size(); i++)
        {
            if(root["type"].asString() == types[i])
            {
                return message;
            }
        }
    }
}

const std::string login_string =
"{\"type\": \"login\", \"args\": {\"username\": \"\", \"connection_type\": \"botnet\"}}";

bool Game::login()
{
    Json::Value event;
    Json::Reader reader_login;
    reader_login.parse(login_string,event,false);

    event["args"]["username"] = ai.username();

    std::stringstream converter;
    std::string login_message;

    //event.asString() does not work
    converter<<event<<std::endl;
    login_message = converter.str();

    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<login_message<<'\n';
    }
    conn.send_string(login_message);

    std::vector<std::string> wanted;
    wanted.push_back("success");
    wanted.push_back("failure");

    std::string message = wait_for(wanted);

    Json::Value root;
    Json::Reader reader;
    reader.parse(message,root,false);
    if(root["type"].asString() == "success")
    {
        return true;
    }
    else
    {
        return false;
    }
}

//create game constant stuff
const std::string create_game_string =
"{\"type\": \"join_game\", \"args\": {}}";

bool Game::create_game()
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(create_game_string,root,false);
    root["args"]["game_name"] = name;

    std::stringstream converter;
    std::string messageSend;
    converter << root;
    messageSend = converter.str();

    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<messageSend<<'\n';
    }
    conn.send_string(messageSend);

    std::vector<std::string> wanted;
    wanted.push_back("success");
    wanted.push_back("failure");

    std::string message = wait_for(wanted);

    Json::Value root2;
    Json::Reader reader2;
    reader.parse(message,root2,false);

    if(root2["type"] == "success")
    {
        name = root2["args"]["name"].asString();
        std::cout<<"Game created: "<<name<<std::endl;
        return true;
    }
    else
    {
        return false;
    }
}

bool Game::recv_player_id()
{
    std::vector<std::string> wanted;
    wanted.push_back("player_id");

    wait_for(wanted);
    return true;
}

bool Game::init_main()
{
    std::vector<std::string> wanted;
    wanted.push_back("start_game");
    wait_for(wanted);

    ai.init();
    return true;
}

bool Game::end_main()
{
    ai.end();
    return true;
}

const std::string end_turn_string =
"{\"type\": \"end_turn\", \"args\": {}}";

bool Game::main_loop()
{
    std::vector<std::string> wanted;
    wanted.push_back("start_turn");
    wanted.push_back("game_over");

    while(true)
    {
        std::string message = wait_for(wanted);
        Json::Value root;
        Json::Reader reader;
        reader.parse(message,root,false);

        if(root["type"] == "game_over")
        {
            return true;
        }

        if(ai.my_player_id == ai.player_id)
        {
            if(verbosity >= 1)
            {
                std::cout<<"Turn Number: "<<ai.turn_number<<std::endl;
            }
            ai.run();
            if(verbosity == 2)
            {
                std::cout<<"Sent: "<<end_turn_string<<'\n';
            }
            conn.send_string(end_turn_string);
        }
    }
}

const std::string get_log_string =
"{\"type\": \"get_log\", \"args\": {}}";

bool Game::get_log()
{
    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<get_log_string<<'\n';
    }
    conn.send_string(get_log_string);

    std::vector<std::string> wanted;
    wanted.push_back("success");
    wanted.push_back("failure");

    std::string message = wait_for(wanted);
    Json::Value root;
    Json::Reader reader;
    reader.parse(message,root,false);

    if(root["type"] == "success")
    {
        std::string fileName = name + ".glog";
        std::ofstream fout;
        fout.open(fileName.c_str());
        fout << root["args"]["log"].asString();
        fout.close();
    }
    return true;
}

bool Game::update_game(std::string message)
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(message,root,false);

    if(root["type"].asString() != "changes")
    {
        return false;
    }

    Json::Value changes = root["args"]["changes"];
    std::stringstream convert;

    for(int i = 0; i < changes.size(); i++)
    {
        convert<<changes[i];
        if(changes[i]["action"].asString() == "add")
        {
            change_add(convert.str());
        }
        else if(changes[i]["action"].asString() == "remove")
        {
            change_remove(convert.str());
        }
        else if(changes[i]["action"].asString() == "update")
        {
            change_update(convert.str());
        }
        else if(changes[i]["action"].asString() == "global_update")
        {
            change_global_update(convert.str());
        }
        convert.str(std::string());
    }

    return true;
}

bool Game::change_add(std::string change)
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(change,root,false);

    Json::Value values = root["values"];
    if(false){}
    else if(root["type"].asString() == "Player")
    {
        Player temp(&conn,this, values["id"].asInt(), values["name"].asString(), values["byte_dollars"].asInt(), values["cycles"].asInt(), values["time"].asInt());
        ai.players.push_back(temp);
    }
    else if(root["type"].asString() == "Base")
    {
        Base temp(&conn,this, values["id"].asInt(), values["x"].asInt(), values["y"].asInt(), values["owner"].asInt(), values["spawns_left"].asInt());
        ai.bases.push_back(temp);
    }
    else if(root["type"].asString() == "Virus")
    {
        Virus temp(&conn,this, values["id"].asInt(), values["x"].asInt(), values["y"].asInt(), values["owner"].asInt(), values["level"].asInt(), values["moves_left"].asInt(), values["living"].asInt());
        ai.viruses.push_back(temp);
    }
    else if(root["type"].asString() == "Tile")
    {
        Tile temp(&conn,this, values["id"].asInt(), values["x"].asInt(), values["y"].asInt(), values["owner"].asInt());
        ai.tiles.push_back(temp);
    }
    else
    {
        std::cout<<"Unknown model attempted to add!!"<<std::endl;
    }
    return true;
}

bool Game::change_remove(std::string change)
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(change,root,false);
    Json::Value null_value;

    int change_id = root["id"].asInt();


    for(int i = 0;i < ai.players.size();i++)
    {
        if(ai.players[i].id == change_id)
        {
            ai.players.erase(ai.players.begin()+i);
            return true;
        }
    }

    for(int i = 0;i < ai.bases.size();i++)
    {
        if(ai.bases[i].id == change_id)
        {
            ai.bases.erase(ai.bases.begin()+i);
            return true;
        }
    }

    for(int i = 0;i < ai.viruses.size();i++)
    {
        if(ai.viruses[i].id == change_id)
        {
            ai.viruses.erase(ai.viruses.begin()+i);
            return true;
        }
    }

    for(int i = 0;i < ai.tiles.size();i++)
    {
        if(ai.tiles[i].id == change_id)
        {
            ai.tiles.erase(ai.tiles.begin()+i);
            return true;
        }
    }
    return false;
}

bool Game::change_update(std::string change)
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(change,root,false);
    Json::Value null_value;

    int change_id = root["id"].asInt();


    for(int i = 0;i < ai.players.size();i++)
    {
        if(ai.players[i].id == change_id)
        {
            if(root["changes"]["id"] != null_value)
            {
                ai.players[i].id = root["changes"]["id"].asInt();
            }
            if(root["changes"]["name"] != null_value)
            {
                ai.players[i].name = root["changes"]["name"].asString();
            }
            if(root["changes"]["byte_dollars"] != null_value)
            {
                ai.players[i].byte_dollars = root["changes"]["byte_dollars"].asInt();
            }
            if(root["changes"]["cycles"] != null_value)
            {
                ai.players[i].cycles = root["changes"]["cycles"].asInt();
            }
            if(root["changes"]["time"] != null_value)
            {
                ai.players[i].time = root["changes"]["time"].asInt();
            }
            return true;
        }
    }

    for(int i = 0;i < ai.bases.size();i++)
    {
        if(ai.bases[i].id == change_id)
        {
            if(root["changes"]["id"] != null_value)
            {
                ai.bases[i].id = root["changes"]["id"].asInt();
            }
            if(root["changes"]["x"] != null_value)
            {
                ai.bases[i].x = root["changes"]["x"].asInt();
            }
            if(root["changes"]["y"] != null_value)
            {
                ai.bases[i].y = root["changes"]["y"].asInt();
            }
            if(root["changes"]["owner"] != null_value)
            {
                ai.bases[i].owner = root["changes"]["owner"].asInt();
            }
            if(root["changes"]["spawns_left"] != null_value)
            {
                ai.bases[i].spawns_left = root["changes"]["spawns_left"].asInt();
            }
            return true;
        }
    }

    for(int i = 0;i < ai.viruses.size();i++)
    {
        if(ai.viruses[i].id == change_id)
        {
            if(root["changes"]["id"] != null_value)
            {
                ai.viruses[i].id = root["changes"]["id"].asInt();
            }
            if(root["changes"]["x"] != null_value)
            {
                ai.viruses[i].x = root["changes"]["x"].asInt();
            }
            if(root["changes"]["y"] != null_value)
            {
                ai.viruses[i].y = root["changes"]["y"].asInt();
            }
            if(root["changes"]["owner"] != null_value)
            {
                ai.viruses[i].owner = root["changes"]["owner"].asInt();
            }
            if(root["changes"]["level"] != null_value)
            {
                ai.viruses[i].level = root["changes"]["level"].asInt();
            }
            if(root["changes"]["moves_left"] != null_value)
            {
                ai.viruses[i].moves_left = root["changes"]["moves_left"].asInt();
            }
            if(root["changes"]["living"] != null_value)
            {
                ai.viruses[i].living = root["changes"]["living"].asInt();
            }
            return true;
        }
    }

    for(int i = 0;i < ai.tiles.size();i++)
    {
        if(ai.tiles[i].id == change_id)
        {
            if(root["changes"]["id"] != null_value)
            {
                ai.tiles[i].id = root["changes"]["id"].asInt();
            }
            if(root["changes"]["x"] != null_value)
            {
                ai.tiles[i].x = root["changes"]["x"].asInt();
            }
            if(root["changes"]["y"] != null_value)
            {
                ai.tiles[i].y = root["changes"]["y"].asInt();
            }
            if(root["changes"]["owner"] != null_value)
            {
                ai.tiles[i].owner = root["changes"]["owner"].asInt();
            }
            return true;
        }
    }
    return false;
}

bool Game::change_global_update(std::string change)
{
    Json::Value root;
    Json::Reader reader;
    reader.parse(change,root,false);
    Json::Value null_value;

    if(root["values"]["turn_number"] != null_value)
    {
        ai.turn_number = root["values"]["turn_number"].asInt();
    }
    if(root["values"]["player_id"] != null_value)
    {
        ai.player_id = root["values"]["player_id"].asInt();
    }
    if(root["values"]["game_number"] != null_value)
    {
        ai.game_number = root["values"]["game_number"].asInt();
    }
    if(root["values"]["base_cost"] != null_value)
    {
        ai.base_cost = root["values"]["base_cost"].asInt();
    }
    if(root["values"]["scale_cost"] != null_value)
    {
        ai.scale_cost = root["values"]["scale_cost"].asFloat();
    }
    if(root["values"]["width"] != null_value)
    {
        ai.width = root["values"]["width"].asInt();
    }
    if(root["values"]["height"] != null_value)
    {
        ai.height = root["values"]["height"].asInt();
    }
    return true;
}

bool Game::run()
{
    std::string game_over_message;
    if(!connect())
    {
        return false;
    }
    if(!login())
    {
        return false;
    }
    if(!create_game())
    {
        return false;
    }
    if(!recv_player_id())
    {
        return false;
    }
    if(!init_main())
    {
        return false;
    }

    try
    {
        main_loop();
    }
    catch(GameOverException e)
    {
        if(e.winner == ai.my_player_id)
        {
            game_over_message = "You Win! - " + e.reason;
        }
        else
        {
            game_over_message = "You Lose! - " + e.reason;
        }
    }
    catch(...)
    {
        game_over_message = "Game over was never reached.";
    }

    if(!end_main())
    {
        return false;
    }

    std::cout<<game_over_message<<std::endl;

    if(!get_log())
    {
        return false;
    }
    return true;
}
